src/node/internal/internal_http_server.ts

```diff
@@ -0,0 +1,735 @@
+    // TODO(soon): Revisit this once we implement net.Server
+    return this;
+  }
+
+  unref(): this {
+    // This method is originally implemented in net.Server.
+    // Since we don't implement net.Server yet, we provide this stub implementation for now.
+    // TODO(soon): Revisit this once we implement net.Server
+    return this;
+  }
+}
+
+// We use this handler to not expose this.#fetchResponse to outside world.
+let getServerResponseFetchResponse: (
+  response: ServerResponse
+) => Promise<Response>;
+
+export class ServerResponse<Req extends IncomingMessage = IncomingMessage>
+  extends OutgoingMessage
+  implements _ServerResponse
+{
+  _sent100 = false;
+  _expect_continue = false;
+  override [kOutHeaders]: Record<string, [string, string | string[]]> | null =
+    null;
+
+  statusCode = 200;
+  statusMessage = 'unknown';
+
+  #fetchResponse: Promise<Response>;
+  #encoder = new TextEncoder();
+
+  static {
+    getServerResponseFetchResponse = (
+      response: ServerResponse
+    ): Promise<Response> => {
+      return response.#fetchResponse;
+    };
+  }
+
+  constructor(req: Req, options: ServerOptions = {}) {
+    super(req, options);
+
+    if (req.httpVersionMajor < 1 || req.httpVersionMinor < 1) {
+      this.useChunkedEncodingByDefault = chunkExpression.test(
+        (req.headers.te as string | undefined) ?? ''
+      );
+      this.shouldKeepAlive = false;
+    }
+
+    const { promise, resolve } = Promise.withResolvers<Response>();
+
+    let finished = false;
+    this.once('finish', () => (finished = true));
+    const chunks: (Buffer | Uint8Array)[] = [];
+    const handler = (event: DataWrittenEvent): void => {
+      if (finished) return;
+      chunks[event.index] = this.#dataFromDataWrittenEvent(event);
+    };
+    this.on('_dataWritten', handler);
+    this.on(
+      '_headersSent',
+      ({ statusCode, statusMessage, headers }: HeadersSentEvent) => {
+        this.off('_dataWritten', handler);
+        resolve(
+          this.#toFetchResponse({
+            statusCode,
+            statusText: statusMessage,
+            sentHeaders: headers,
+            chunks,
+            finished,
+          })
+        );
+      }
+    );
+    this.#fetchResponse = promise;
+  }
+
+  #toFetchResponse({
+    statusCode,
+    statusText,
+    sentHeaders,
+    chunks,
+    finished,
+  }: {
+    statusCode: number;
+    statusText: string;
+    sentHeaders: [header: string, value: string][];
+    chunks: (Buffer | Uint8Array)[];
+    finished: boolean;
+  }): Response {
+    const headers = new Headers();
+    for (const [header, value] of sentHeaders) {
+      headers.append(header, value);
+    }
+
+    let body = null;
+
+    if (this._hasBody) {
+      const _this = this; // eslint-disable-line @typescript-eslint/no-this-alias
+      body = new ReadableStream<Uint8Array>({
+        start(controller): void {
+          for (const chunk of chunks) {
+            controller.enqueue(chunk);
+          }
```

Comment:
Hmm.. this seems, inefficient. We ought to be able to stream directly from the Node.js OutgoingMessage stream into a the ReadableStream. As this is written, this is going to end up buffering the data twice is it not? Once in the OutgoingMessage and then in the ReadableStream passed to the `Response` object?

Reply:
still, before this lands let's make sure there's a test that writes a large response payload over multiple calls to ensure that everything is streaming the way it should.